pkg/types/container/v0.0.1/entry.go

```diff
@@ -0,0 +1,392 @@
+	"github.com/pkg/errors"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/pki"
+	"github.com/sigstore/rekor/pkg/types"
+	"github.com/sigstore/rekor/pkg/types/container"
+)
+
+const (
+	APIVERSION = "0.0.1"
+)
+
+func init() {
+	if err := container.VersionMap.SetEntryFactory(APIVERSION, NewEntry); err != nil {
+		log.Logger.Panic(err)
+	}
+}
+
+type V001Entry struct {
+	ContainerObj models.ContainerV001Schema
+}
+
+func (v V001Entry) APIVersion() string {
+	return APIVERSION
+}
+
+func NewEntry() types.EntryImpl {
+	return &V001Entry{}
+}
+
+func (v V001Entry) IndexKeys() ([]string, error) {
+	var result []string
+
+	// TODO: Do we always expect x509 keys signing containers?
+	keyObj, err := x509.NewPublicKey(bytes.NewReader(v.ContainerObj.Signature.PublicKey.Content))
+	if err != nil {
+		return nil, err
+	}
+
+	key, err := keyObj.CanonicalValue()
+	if err != nil {
+		log.Logger.Error(err)
+	} else {
+		keyHash := sha256.Sum256(key)
+		result = append(result, strings.ToLower(hex.EncodeToString(keyHash[:])))
+	}
+
+	result = append(result, keyObj.EmailAddresses()...)
+
+	// This is the hash of the signed payload.
+	if v.ContainerObj.Data.Hash != nil {
+		hashKey := strings.ToLower(fmt.Sprintf("%s:%s", *v.ContainerObj.Data.Hash.Algorithm, *v.ContainerObj.Data.Hash.Value))
+		result = append(result, hashKey)
+	}
+
+	// This is the container digest that is signed.
+	if v.ContainerObj.Data.Content != nil {
+		// Marshal and unmarshal into a payload.Cosign.
+		b, err := json.Marshal(v.ContainerObj.Data.Content)
+		if err != nil {
+			log.Logger.Error(err)
+		} else {
+			img := &payload.Cosign{}
+			if err := json.Unmarshal(b, img); err == nil {
+				result = append(result, img.Image.DigestStr())
+			}
+		}
+		// TODO: Should we include the image reference as well?
+		// TODO: What to do about optional annotations?
+	}
+
+	return result, nil
+}
+
+func (v *V001Entry) Unmarshal(pe models.ProposedEntry) error {
+	container, ok := pe.(*models.Container)
+	if !ok {
+		return errors.New("cannot unmarshal non container v0.0.1 type")
+	}
+
+	if err := types.DecodeEntry(container.Spec, &v.ContainerObj); err != nil {
+		return err
+	}
+
+	// field validation
+	if err := v.ContainerObj.Validate(strfmt.Default); err != nil {
+		return err
+	}
+
+	// cross field validation
+	return v.validate()
+
+}
+
+func (v *V001Entry) hasExternalEntities() bool {
+	if v.ContainerObj.Data != nil && v.ContainerObj.Data.URL.String() != "" {
+		return true
+	}
+	// TODO: Allow signature/public key URLs or no?
+	return false
+}
+
+func (v *V001Entry) fetchExternalEntities(ctx context.Context) (pki.PublicKey, pki.Signature, error) {
+	oldSHA := ""
+	if v.ContainerObj.Data.Hash != nil && v.ContainerObj.Data.Hash.Value != nil {
+		oldSHA = swag.StringValue(v.ContainerObj.Data.Hash.Value)
+	}
+
+	var contentBytes []byte
+	if v.ContainerObj.Data.Content != nil {
+		var err error
+		contentBytes, err = json.Marshal(v.ContainerObj.Data.Content)
```

Comment:
what happens if this marshalling ends up with the elements in a different order than when the SHA value (also in the same input doc) was computed?

Reply:
Hmmmm this is a good point. Ideally I think this should compare the SHA of the original `props.ArtifactBytes` that's originally passed in to avoid the unmarshal/marshal cycle.